avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java

```diff
@@ -492,21 +494,41 @@ private void processUnion(JVar unionSchemaVar, final String name, final Schema u
       final Schema readerUnionSchema, JBlock body, FieldAction action,
       BiConsumer<JBlock, JExpression> putValueIntoParent, Supplier<JExpression> reuseSupplier) {
     JVar unionIndex = body.decl(codeModel.INT, getUniqueName("unionIndex"), JExpr.direct(DECODER + ".readIndex()"));
-    JConditional ifBlock = null;
+    JSwitch switchBlock = body._switch(unionIndex);
     for (int i = 0; i < unionSchema.getTypes().size(); i++) {
       Schema optionSchema = unionSchema.getTypes().get(i);
       Schema readerOptionSchema = null;
+      int readerOptionUnionBranchIndex = -1;
       FieldAction unionAction;
-      JExpression condition = unionIndex.eq(JExpr.lit(i));
+      JBlock caseBody = switchBlock._case(JExpr.lit(i)).body();
 
       if (Schema.Type.NULL.equals(optionSchema.getType())) {
-        ifBlock = ifBlock != null ? ifBlock._elseif(condition) : body._if(condition);
-        ifBlock._then().directStatement(DECODER + ".readNull();");
+        caseBody.directStatement(DECODER + ".readNull();");
+        caseBody._break();
         continue;
       }
 
       if (action.getShouldRead()) {
-        readerOptionSchema = readerUnionSchema.getTypes().get(i);
+        // The reader's union could be re-ordered, so we need to find the one that matches.
+        for (int j = 0; j < readerUnionSchema.getTypes().size(); j++) { //  Schema potentialReaderSchema: readerUnionSchema.getTypes()) {
+          Schema potentialReaderSchema = readerUnionSchema.getTypes().get(j);
+          // Avro allows unnamed types to appear only once in a union, but named types may appear multiple times and
+          // thus need to be disambiguated via their full-name (including aliases).
+          if (potentialReaderSchema.getType().equals(optionSchema.getType()) &&
+              (!schemaAssistant.isNamedType(potentialReaderSchema) ||
+                  potentialReaderSchema.getFullName().equals(optionSchema.getFullName()) ||
+                  potentialReaderSchema.getAliases().contains(optionSchema.getFullName()))) {
+            readerOptionSchema = potentialReaderSchema;
+            readerOptionUnionBranchIndex = j;
```

Comment:
is it worth completing the loop to make sure the result is unambiguous ?

Reply:
IDK... I guess it should be illegal to have multiple union branches with the same full name. We could construct a test that checks what is the Avro behavior in such cases and emulate the same behavior.